### PR TITLE
Auth/PM-1658 - Dynamic Org Invite Link to accelerate users through org invite accept process 

### PR DESCRIPTION
**Pull request by JaredSnider-Bitwarden**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

[PM-1658](https://bitwarden.atlassian.net/browse/PM-1658)

Associated server PR here: https://github.com/bitwarden/server/pull/3378

This changes the place in the application users are taken when accepting an organization invite, accelerating them through the process. 

1. Existing users will be accelerated to the login screen. 
2. New users in an Org with SSO enabled and the SSO Login required policy enabled will be taken to the start of the SSO step with the Org SSO Identifier filled in for them
3. New users in an Org with SSO enabled with the SSO Login required policy disabled or in an Org with SSO disabled will be taken to the register screen with the email readonly

## Code changes

- **apps/web/src/app/auth/accept-organization.component.ts:** Update accept org component to add a new `accelerateInviteAcceptIfPossible(...)` method to the `unauthedHandler(...)` based on the presence of new query params sent down 
- **apps/web/src/app/auth/trial-initiation/trial-initiation.component.ts** Add new `fromOrgInvite` query param handling 
- **apps/web/src/app/auth/trial-initiation/trial-initiation.component.html**  Pass `fromOrgInvite` query param value into nested `register-form` child component
- **apps/web/src/app/auth/register-form/register-form.component.ts** Add new input property of `queryParamFromOrgInvite`
- **apps/web/src/app/auth/register-form/register-form.component.html** Use new `queryParamFromOrgInvite` property to make email readonly when the query param is present. 

## Screenshots

Before these changes, users would land on the accept-organization screen to choose login or create account (this flow still exists as a fallback if any of the new query params are missing):
image.

Existing User taken to login:

https://github.com/bitwarden/clients/assets/116684653/777a4ba1-8eaf-446e-b928-767ab5399774

New User joining Org with SSO enabled and required by policy taken to start of SSO flow:

https://github.com/bitwarden/clients/assets/116684653/1958340f-f678-4709-ad44-fe3a553010ff

New User joining Org with SSO enabled but not required by policy taken to register with non-editable email:
https://github.com/bitwarden/clients/assets/116684653/11822fd6-81d6-442e-932c-3d9de848491e

New User joining Org with SSO disabled taken to register with non-editable email:

https://github.com/bitwarden/clients/assets/116684653/5973f589-a72c-46ab-841a-1833b90bc69a

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

[PM-1658]: https://bitwarden.atlassian.net/browse/PM-1658?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/web/src/app/auth/accept-organization.component.ts [hunk 1]
```diff
@@ -186,4 +190,54 @@ export class AcceptOrganizationComponent extends BaseAcceptComponent {
...
+      orgUserHasExistingUser == null
+    ) {
+      return;
+    }
+
+    // if user exists, send user to login
+    if (orgUserHasExistingUser) {
+      this.router.navigate(["/login"]);
+      return;
+    }
```
Comment (jlf0dev) [inline apps/web/src/app/auth/accept-organization.component.ts:217]: Does this autofill the email? Do we need to pass that as a query param?

File: apps/web/src/app/auth/accept-organization.component.ts [hunk 2]
```diff
@@ -186,4 +190,54 @@ export class AcceptOrganizationComponent extends BaseAcceptComponent {
...
+
+    // no user exists; so either register or sign in via SSO and JIT provision one.
+    // if SSO is disabled OR if sso is enabled but the SSO login required policy is not enabled
```
Comment (jlf0dev) [inline apps/web/src/app/auth/accept-organization.component.ts:220]: Made a comment on server PR as well, but we should check with product if this is required. I don't see why we need the policy to be sent down, if you have SSO enabled you should just be sent to `/sso`. Then we could send a single parameter for `ssoIdentifier`.

Comment (JaredSnider-Bitwarden) [reply apps/web/src/app/auth/accept-organization.component.ts:220]: This logic came out of a direct conversation with product. They only wanted to force the user through the SSO JIT provisioning process if the require SSO policy was enabled.

File: apps/web/src/app/auth/accept-organization.component.ts (hunk 1 above)
Comment (JaredSnider-Bitwarden) [reply apps/web/src/app/auth/accept-organization.component.ts:217]: The email information is saved in state in the `prepareOrganizationInvitation(...)` method in the `stateService.setOrganizationInvitation(...)` call.  Then, the login component leverages `stateService.getOrganizationInvitation(...)` to pre-populate the email on load of the login component.

File: apps/web/src/app/auth/accept-organization.component.ts (hunk 2 above)
Comment (JaredSnider-Bitwarden) [reply apps/web/src/app/auth/accept-organization.component.ts:220]: Simplified! Thank you!

**Review by jlf0dev — DISMISSED** [review]
Looks great!

**Comment by JaredSnider-Bitwarden** [comment]
Passed QA. Merging after builds & tests pass.

**Pull request merged**